packages/hub/routes/status.ts

```diff
@@ -0,0 +1,67 @@
+import Koa from 'koa';
+import autoBind from 'auto-bind';
+import { inject } from '@cardstack/di';
+import * as Sentry from '@sentry/node';
+
+const DEGRADED_THRESHOLD = 10;
+
+export default class StatusRoute {
+  subgraph = inject('subgraph');
+  web3 = inject('web3-http', { as: 'web3' });
+
+  constructor() {
+    autoBind(this);
+  }
+
+  async get(ctx: Koa.Context) {
+    let subgraphBlockNumber = null;
+
+    try {
+      let subgraphMeta = await this.subgraph.getMeta();
+      subgraphBlockNumber = subgraphMeta?.data?._meta?.block?.number;
+    } catch (e) {
+      Sentry.captureException(e, {
+        tags: {
+          action: 'status-route',
+        },
+      });
+    }
+
+    let rpcBlockNumber = null;
+
+    try {
+      rpcBlockNumber = await this.web3.getInstance().eth.getBlockNumber();
+    } catch (e) {
+      Sentry.captureException(e, {
+        tags: {
+          action: 'status-route',
+        },
+      });
+    }
+
+    let status = 'degraded';
+
+    if (rpcBlockNumber && subgraphBlockNumber && rpcBlockNumber - subgraphBlockNumber < DEGRADED_THRESHOLD) {
+      status = 'operational';
+    }
+
+    ctx.status = 200;
+    ctx.body = {
+      data: {
+        type: 'status',
+        attributes: {
+          status,
```

Comment:
Would it be good to include the degradation notice title and message? Otherwise I imagine it would take some duplicated logic to form a meaningful degradation message across the apps that will consume this endpoint (dApp, Card Wallet).

Am I right to anticipate the /status endpoint could contain more metrics in the future? In this case I think it would be good to include info on which metric is the culprit for the degradation.

Reply:
There's also a task where we'd like to have an externally hosted status page. I think those kinds of pages usually show the status of different services. Would it be good to model our status endpoint to include the status of individual services as well? Maybe something like:

```
status: {
 subgraph: {
   status: "degraded",
   reason: "",
  subgraphBlockNumber,
  rpcBlockNumber
 },
 someOtherService: {
   status: "operational",
   ... 
 }, 
 {...}
}
```

Reply:
Great suggestions, thanks. I’ve updated to move all subgraph status-related information into a `subgraph` container. I’ve also added an `errors` top-level container when there’s an error that has a relevant `title` and a `source`.

The known failure of the subgraph being behind seems self-evident:

```
{
        data: {
          type: 'status',
          attributes: {
            subgraph: {
              rpcBlockNumber: 19492430,
              status: 'degraded',
              subgraphBlockNumber: 19492419,
            },
          },
        },
        errors: [
          {
            id: 'subgraph',
            title: 'Experiencing slow service',
            source: {
              pointer: '/data/attributes/subgraph/subgraphBlockNumber',
            },
          },
        ],
      }
```

However, representing a failure to query the subgraph or the RPC node is more challenging as the [error object specification](https://jsonapi.org/format/#error-objects) only supports a `parameter` or `pointer` under `source`; I used `service` but maybe the `title` alone is sufficient?
```
{
    data: {
        type: 'status',
        attributes: {
        subgraph: {
            rpcBlockNumber: null,
            status: 'degraded',
            subgraphBlockNumber: 19492428,
        },
        },
    },
    errors: [
        {
        id: 'subgraph',
        title: 'Error checking status',
        source: {
            service: 'web3-http',
        },
        },
    ],
}
```

Reply:
Cool, thanks for considering this!. Do you think we can drop the `errors` part and move the explanation to the `attributes`? I feel like the `errors` is more suitable for model validation errors while here we're just serving some information. My suggestion:

```
data: {
  type: 'status',
  attributes: {
    subgraph: {
      status: 'degraded',
      rpcBlockNumber: 19492430,
      subgraphBlockNumber: 19492419,
      details: "Experiencing slow service"
    },
  },
},
```

And in cases when there is a failure to query, I'd consider setting the status to `unknown` 🤷‍♂️

Reply:
ya, makes sense that it isn’t really fitting with errors in JSON:API. I’ve done both of those now, thanks!